### PR TITLE
Fixing connect/disconnect and adding UI elements - connect/disconnect buttons and status indicator

### PR DESCRIPTION
**Pull request by mspnr**
This PR addresses the issue where OctoPrint-MQTT fails to reconnect after the MQTT broker (Home Assistant) restarts.  See #151 

_Note: It was also requested to merge with devel branch, but it is 14 commit behind master, so I am not sure it makes sense._

# Bug Fixes

## 1. Fixed MQTT disconnect/reconnect logic
- **Problem**: The `mqtt_disconnect()` method never called `self._mqtt.disconnect()`, so the broker didn't receive a      
proper disconnect packet. This caused reconnection attempts to fail because the broker thought the client was still       
connected.
- **Solution**:
- Added `self._mqtt.disconnect()` call before stopping the network loop

## 2. Fixed invalid topic subscription error
- **Problem**: When reconnecting, the plugin attempted to subscribe to topics from the subscription list, but some        
entries had invalid/empty topic values, causing a `ValueError: Invalid topic` exception. This prevented
`_mqtt_connected` from being set to `True`.
- **Solution**:
- Added filtering to exclude invalid topics (None or empty strings) before subscribing
- Added exception handling in `_on_mqtt_connect()` to log errors without breaking the connection

# New Features

## Added manual Connect/Disconnect controls in settings UI
- Added **Connection Status** indicator showing real-time connection state (Connected/Disconnected)
- Added **Disconnect** button (red) to manually disconnect from broker
- Added **Connect** button (green) to manually reconnect to broker
- Implemented `SimpleApiPlugin` with admin-only API endpoints:
- `POST /api/plugin/mqtt` with command `connect` - initiates connection
- `POST /api/plugin/mqtt` with command `disconnect` - disconnects from broker
- `GET /api/plugin/mqtt` - returns current connection status
- Connection status uses `is_connected()` from Paho MQTT client for accurate real-time status

# Benefits
- Users can now manually reconnect to MQTT broker after Home Assistant restarts without restarting OctoPrint
- Visual feedback shows current connection state
- Proper disconnect/reconnect sequence ensures clean broker sessions
- Admin-only API access ensures security

# Screenshots

image

image

### CODE REVIEW

File: octoprint_mqtt/__init__.py [hunk 1]
```diff
@@ -349,6 +351,55 @@ def get_update_information(self):
...
+            try:
+                actual_connected = self._mqtt.is_connected()
+                # Sync our flag with actual status
+                if actual_connected != self._mqtt_connected:
+                    self._logger.debug("Connection status out of sync, updating from {} to {}".format(
+                        self._mqtt_connected, actual_connected))
+                    self._mqtt_connected = actual_connected
+            except:
+                # is_connected() might not be available in older paho versions
+                actual_connected = self._mqtt_connected
```
Comment (jneilliii) [inline octoprint_mqtt/__init__.py]: might make more sense to pin the dependency to minimum version 1.4.0 that supports is_connected() in `plugin_requires` of setup.py. That's the version this method was added to, which was released on September 2, 2018.

File: octoprint_mqtt/__init__.py [hunk 2]
```diff
@@ -417,6 +468,12 @@ def mqtt_disconnect(self, force=False, incl_lwt=True, lwt=None):
                 _retain = self._settings.get_boolean(["broker", "lwRetain"])
                 self._mqtt.publish(lwt, self.LWT_DISCONNECTED, qos=1, retain=_retain)
 
+        # Actually disconnect from the broker
+        try:
+            self._mqtt.disconnect()
+        except:
+            pass  # Ignore errors if already disconnected
```
Comment (jneilliii) [inline octoprint_mqtt/__init__.py]: Assuming we pin the version in the comment above it would make more sense to check if_connected()? 

```py
if self._mqtt.is_connected():
    self._mqtt.disconnect()
```

**Review by jneilliii — COMMENTED** [review]
Thanks for your contribution, just a couple of code comments.

File: octoprint_mqtt/__init__.py (hunk 1 above)
Comment (jneilliii) [reply octoprint_mqtt/__init__.py]: might even make more sense to update the dependency to minimum version 2 and adjust the code accordingly elsewhere to support that change. pinning to less than 2 was just a quick fix due to breaking changes. only reason I say that is because of the current [build status of piwheels for this dependency](https://www.piwheels.org/project/paho-mqtt/), which you can see by expanding releases, only the 2.0 builds are available for both major python version 3.9 and 3.11.

Comment (mspnr) [reply octoprint_mqtt/__init__.py]: I see that there are some breaking changes in paho-mqtt in version 2. I would proceed with smaller steps: at first pin `paho-mqtt>=1.4.0,<2`, merge the change and then upgrade to version 2 in a separate pull request.

File: octoprint_mqtt/__init__.py (hunk 2 above)
Comment (mspnr) [reply octoprint_mqtt/__init__.py]: Adding check `is_connected()` make sense. 

I will do the change as soon as my long print will be finished tomorrow, so I can test the changes. I will post the changes here.

**Comment by mspnr** [comment]
@jneilliii added [200d681](https://github.com/OctoPrint/OctoPrint-MQTT/pull/155/commits/200d681e5c6b52fbe5f099566c328c3c7a80a432) according to your suggestions:

## Changes Made

Thank you for the review feedback! I've addressed both suggestions:

### 1. Dependency Management
Pinned minimum paho-mqtt version to **1.4.0** (when `is_connected()` was introduced)
- Updated `setup.py`: `"paho-mqtt>=1.4.0,<2"`
- This ensures the method is always available without needing fallback logic

### 2. Proper Use of `is_connected()`
Removed bare exception handling and replaced with explicit connection checks:

**In `on_api_get()` (lines 386-397):**
- Removed try/except wrapper around `is_connected()` call
- No longer needed since we guarantee paho-mqtt >= 1.4.0

**In `mqtt_disconnect()` (lines 456-476):**
- Replaced `try: self._mqtt.disconnect() except: pass` with `if self._mqtt.is_connected(): self._mqtt.disconnect()`     
- Added connection check before publishing LWT message
- Prevents unnecessary operations when already disconnected

### Why Option A (1.4.0+) vs Option B (2.x)?
I chose the conservative approach to maintain **Python 2.7+ compatibility** as specified in 
`__plugin_pythoncompat__`. If you prefer upgrading to paho-mqtt 2.x, I can update the Python compatibility 
requirement as well.

These changes eliminate bare exception handling while maintaining backward compatibility. Ready for another review!

**Comment by jneilliii** [comment]
@Kunsi I don't see any other code issues related to this PR, but haven't tested yet. I feel like there could be some changes to how the connect/disconnect buttons and connection status area are handled to utilize ko bindings instead of DOM manipulation via js, but technically should still work as-is in this PR.

**Comment by mspnr** [comment]
# Testing and bonus effects

Every time before pushing commits I test the function on my instance of OctoPrint.

Moreover - after making these changes - OctoPrint keeps connection even after restart of Home Assistant, issue described here #151 

# DOM manipulation vs ko bindings

As I understood you have concern about this commit https://github.com/OctoPrint/OctoPrint-MQTT/pull/155/commits/2a74c82d3115862685428cb21f8314cd0fc49833 . Fair enough - I improved the situation here: https://github.com/OctoPrint/OctoPrint-MQTT/pull/155/commits/70514e372d3c1fbe3b5a93ac4c9dc430483fa889

I've refactored the connect/disconnect buttons and connection status to use ko.observables and computed observables     
instead of jQuery DOM manipulation. The changes include:

- Added observables: `isConnected`, `isConnecting`, `isDisconnecting`
- Created computed observables for status HTML and button states
- Removed all `$("#...")` selectors and direct DOM manipulation
- Updated template to use `data-bind` attributes

The UI now follows OctoPrint's standard ViewModel patterns and is more maintainable. All functionality remains the      
same - just implemented properly with reactive bindings.

**Comment by jneilliii** [comment]
I made some adjustments to AI's disable button observables to include isConnected() checks in https://github.com/mspnr/OctoPrint-MQTT/commit/b3802b8597850a4938088e9c875fbaf944096f5e. One last issue I notice is that the connection state isn't getting updated in the front-end when the back-end loses the connection to the broker. To test this yourself, be in a connected state and stop your MQTT broker. The plugin will lose it's connection but the UI still thinks it's connected. Same for when you start the broker back up and plugin is in disconnected state. 

image

Probably need to add a [self._plugin_manager.send_plugin_message(self._identifier, {"connected": False})](https://docs.octoprint.org/en/main/modules/plugin.html#octoprint.plugin.core.PluginManager.send_plugin_message) to _on_mqtt_disconnect function and `self._plugin_manager.send_plugin_message(self._identifier, {"connected": True})` to the _on_mqtt_connect function and call the newly added updateConnectionStatus with data.connected payload.

**Comment by mspnr** [comment]
Thank you for improving buttons behavior!

I added sync MQTT connection status between backend and frontend and also tested that. It is working just fine. Actually pressing connect is not needed anymore, because my tests showed, that OctoPrint always connected back to the broker as soon as it became available again. I would keep Connect/Disconnect buttons for debugging or testing edge cases in configuration.
